### PR TITLE
[RNMobile] Simplify media insertion flow - Part 3 component integration

### PR DESCRIPTION
**Pull request by jd-alexander**
`gutenberg-mobile` PR https://github.com/wordpress-mobile/gutenberg-mobile/pull/2700

- [Part 1](https://github.com/WordPress/gutenberg/pull/29546)
- [Part 2](https://github.com/WordPress/gutenberg/pull/29547)
- [Part 3](https://github.com/WordPress/gutenberg/pull/29548) ( Current ) 

Builds on top of https://github.com/WordPress/gutenberg/pull/29547 - Do not merge until this PR has been merged and the base branch is updated to `trunk`. 

## Description
This change simplifies the media insertion flow by showing the media options sheet when you tap `Image`, `Video` or `Gallery` to add the media placeholder to the canvas. Before you would have to add the media placeholder and click `Add Image / Video / Media` for it to be shown after insertion. 

## Testing 

1. Tap (+) (inserter) – this shows the Block Library sheet
2. Tap Image (or Video or Gallery) to add the block's media placeholder to canvas – which triggers the media upload options sheet. 
3. Notice the media upload options auto open.

### Image

Android | iOS 
--------|-------
 <kbd></kbd>    |       <kbd></kbd>

### Video

Android | iOS 
--------|-------
  <kbd></kbd>    |     <kbd></kbd>
  
### Gallery

Android | iOS 
--------|-------
<kbd></kbd>    |      <kbd></kbd>
 
----------

### Regression Testing

1. Verify that undo/redo on all media blocks does not trigger the auto-opening of the media options sheet.
2. Verify that selecting the media-related blocks in the canvas does not trigger the auto opening. ( Note : When you select a media-related block on the emulator, the placeholder may receive the touch event and open the media options sheet.)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-library/src/gallery/edit.js [hunk 1]
```diff
@@ -464,6 +468,22 @@ export default compose( [
...
+				const result = wasBlockJustInserted( clientId );
+
+				if ( result ) {
+					clearLastBlockInserted();
+					return true;
+				}
+				return false;
+			},
+		};
+	} ),
```
Comment (jd-alexander) [inline packages/block-library/src/gallery/edit.js:486]: I repeated this `withDispatch` logic in three places. I am not sure how to make this reusable hence the duplication. Instinctively, I thought of hooks but these are class components so they are incompatible and the nature of this PR is not to do this conversion for these components. Let me know if you have any thoughts on this.

File: packages/block-library/src/image/edit.native.js [hunk 2]
```diff
@@ -579,5 +588,21 @@ export default compose( [
 			imageSizes,
 		};
 	} ),
+	withDispatch( ( dispatch, { clientId }, { select } ) => {
+		return {
+			wasBlockJustInserted() {
+				const { clearLastBlockInserted } = dispatch( 'core/editor' );
+				const { wasBlockJustInserted } = select( 'core/editor' );
```
Comment (jd-alexander) [inline packages/block-library/src/image/edit.native.js:595]: I am getting a warning about using the string literals to access the data store. I haven't seen a store definition for `core-editor` in other parts of the codebase as yet.

File: packages/block-library/src/gallery/edit.js (hunk 1 above)
Comment (jd-alexander) [reply packages/block-library/src/gallery/edit.js:486]: I also wondered if this logic could somehow live within the store, but I wasn't sure about this either.

File: packages/react-native-editor/__device-tests__/gutenberg-editor-gallery.test.js [hunk 3]
```diff
@@ -6,6 +6,9 @@ import { blockNames } from './pages/editor-page';
 describe( 'Gutenberg Editor Gallery Block tests', () => {
 	it( 'should be able to add a gallery block', async () => {
 		await editorPage.addNewBlock( blockNames.gallery );
+		await editorPage.driver.sleep( 1000 );
```
Comment (jd-alexander) [inline packages/react-native-editor/__device-tests__/gutenberg-editor-gallery.test.js:9]: This delay and the one below ensures the picker has been opened before attempting to close it.

File: packages/block-library/src/image/edit.native.js (hunk 2 above)
Comment (jd-alexander) [reply packages/block-library/src/image/edit.native.js:595]: I just found the [PR that replaced store names](https://github.com/WordPress/gutenberg/pull/28975) and I saw that the `core/editor` is now `editorStore`

Comment (jd-alexander) [reply packages/block-library/src/image/edit.native.js:595]: I have the changes for this locally. I am encountering the error below. I will look into what could be causing in the morning

<kbd></kbd>

**Review by chipsnyder — COMMENTED** [review]
Nice work @jd-alexander! Tested this out on iOS and Android using the demo apps and everything works as expected. I agree with your comment to reuse the `withDispatch` changes but I've been away from RN so I didn't have a good suggestion there yet. 

One other thought, I wonder if we should auto-open on `File` and `Audio` as well. WDYT?

File: packages/block-library/src/image/edit.native.js (hunk 2 above)
Comment (chipsnyder) [reply packages/block-library/src/image/edit.native.js:595]: I didn't see this error pop up for me so you probably have it fixed now. This sounds similar to an issue solved in [this PR](https://github.com/WordPress/gutenberg/pull/25523) where a setting was removed that the mobile side counted on.

**Comment by jd-alexander** [comment]
> Nice work @jd-alexander! Tested this out on iOS and Android using the demo apps and everything works as expected. I agree with your comment to reuse the withDispatch changes but I've been away from RN so I didn't have a good suggestion there yet.

Thanks for the review 🙇🏾 

> One other thought, I wonder if we should auto-open on File and Audio as well. WDYT?

Good point, I think it would make sense to open them as well. I am going to create another PR with that implementation  once these have been finalized and merged in.

**Comment by guarani** [comment]
BTW, the failing iOS E2E tests might have been fixed by https://github.com/WordPress/gutenberg/pull/29589, so if you update with the latest from `trunk`, the build might pass 🤞

**Review by chipsnyder — APPROVED** [review]
> Good point, I think it would make sense to open them as well. I am going to create another PR with that implementation once these have been finalized and merged in.

This sounds like a good plan to me 👍

**Review by guarani — APPROVED** [review]
Tested on iOS and Android and things are looking good!
The demo app on Android doesn't allow inserting images from the device, so it's hard to test the full user flow here, but I think we're good.

**Comment by jd-alexander** [comment]
> Tested on iOS and Android and things are looking good!
> The demo app on Android doesn't allow inserting images from the device, so it's hard to test the full user flow here, but I think we're good.

Thanks much for all the reviews and comments @guarani 🙇🏾 . I have synced all branches with trunk and I will be resolving the test issues and merging. A label in the e2e test says "Dismiss" instead of "Cancel" so I will update that.

**Pull request merged**